nexus/src/app/instance.rs

```diff
@@ -72,48 +74,82 @@ type SledAgentClientError =
 
 // Newtype wrapper to avoid the orphan type rule.
 #[derive(Debug, thiserror::Error)]
-pub struct SledAgentInstancePutError(pub SledAgentClientError);
+pub struct SledAgentInstanceError(pub SledAgentClientError);
 
-impl std::fmt::Display for SledAgentInstancePutError {
+impl std::fmt::Display for SledAgentInstanceError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(f, "{}", self.0)
     }
 }
 
-impl From<SledAgentClientError> for SledAgentInstancePutError {
+impl From<SledAgentClientError> for SledAgentInstanceError {
     fn from(value: SledAgentClientError) -> Self {
         Self(value)
     }
 }
 
-impl From<SledAgentInstancePutError> for omicron_common::api::external::Error {
-    fn from(value: SledAgentInstancePutError) -> Self {
+impl From<SledAgentInstanceError> for omicron_common::api::external::Error {
+    fn from(value: SledAgentInstanceError) -> Self {
         value.0.into()
     }
 }
 
-impl SledAgentInstancePutError {
-    /// Returns `true` if this error is of a class that indicates that Nexus
-    /// cannot assume anything about the health of the instance or its sled.
+impl From<SledAgentInstanceError> for dropshot::HttpError {
+    fn from(value: SledAgentInstanceError) -> Self {
+        use dropshot::HttpError;
+        // See RFD 486 §6.3:
+        // https://rfd.shared.oxide.computer/rfd/486#_stopping_or_rebooting_a_running_instance
+        match value.0 {
+            // Errors communicating with the sled-agent should return 502 Bad
+            // Gateway to the caller.
+            progenitor_client::Error::CommunicationError(e) => {
+                // N.B.: it sure *seems* like we should be using the
+                // `HttpError::for_status` constructor here, but that
+                // constructor secretly calls the `for_client_error`
+                // constructor, which panics if we pass a status code that isn't
+                // a 4xx error. So, instead, we construct an internal error and
+                // then munge its status code.
+                let mut error = HttpError::for_internal_error(e.to_string());
+                error.status_code = if e.is_timeout() {
+                    // This isn't actually in the RFD, but I thought it might be
+                    // nice to return 504 Gateway Timeout when the communication
+                    // error is a timeout...
+                    http::StatusCode::GATEWAY_TIMEOUT
+                } else {
+                    http::StatusCode::BAD_GATEWAY
+                };
+                error
+            }
+            // Invalid request errors from the sled-agent should return 500
+            // Internal Server Errors.
+            progenitor_client::Error::InvalidRequest(s) => {
+                HttpError::for_internal_error(s)
+            }
+            e => HttpError::for_internal_error(e.to_string()),
+        }
```

Comment:
@gjcolombo do you think it's worth adding an external message to any of these errors? I was thinking that if `self.instance_unhealthy()` is `true`, it might be nice to return an error with a message like "instance failure detected" (and perhaps that should be a 503 Service Unavailable rather than 500?), and maybe the 502s and 504s should say "error communicating with sled-agent" or something? I dunno...